feat!: add concurrent processing and migrate to loguru for enhanced performance and reliability

kWAYTV opened the pull request:
## Summary
Enhances the CS Medal Parser with significant performance and reliability improvements while migrating to modern logging.

## Key Changes
- **Performance**: Concurrent downloads with 10 parallel workers (~10x faster image processing)
- **Reliability**: HTTP connection pooling, retry logic, and timeout management  
- **Logging**: Migrated from standard logging to loguru for better structured output
- **Configuration**: Added new collectible types "pass" and "stars"
- **Docker**: Updated base image to Python 3.13.5-alpine

## Technical Improvements
- ThreadPoolExecutor for parallel image downloads
- Automatic retry strategy for failed network requests
- Enhanced error handling and image validation
- Real-time progress tracking and performance metrics
- Optimized regex patterns for faster filtering

## Breaking Changes
- Requires `loguru` dependency
- Logging configuration removed from config.py (uses loguru defaults)

## Impact
- ~10x faster image processing for large batches
- Significantly more reliable network operations
- Better debugging experience with structured logging
- Production-ready error handling

All existing functionality preserved while adding new capabilities.

kWAYTV committed:
feat(config): add new collectible types "pass" and "stars"

kWAYTV committed:
feat(logging): enhance logging and error handling in main and config

- Added logging configuration to main and test scripts.
- Improved error handling in fetch_collectibles and dump_collectibles functions.
- Introduced concurrent image downloading with progress tracking.
- Added new configuration options for performance and image processing in config.py.

kWAYTV committed:
chore(config): remove logging configuration from config.py and update README

- Removed logging configuration from config.py.
- Updated README to reflect new features and installation instructions.
- Added loguru as a dependency in requirements.txt for enhanced logging.

kWAYTV committed:
chore(docker): update Python base image to 3.13.5-alpine

kWAYTV committed:
chore(gitignore): add filtered.json to .gitignore

kWAYTV committed:
feat(parser): refactor CSMedalParser for improved collectible processing

- Introduced CSMedalParser class for handling the parsing and processing of CS:GO collectibles.
- Enhanced error handling and logging throughout the parsing process.
- Added support for concurrent image downloading and processing.
- Updated main function to utilize the new parser structure.
- Refactored existing functions for better organization and maintainability.
- Updated requirements.txt to include aiohttp for asynchronous operations.

kWAYTV committed:
style: format import statements for consistency

- Reformatted import statements in main.py, parser.py, api_service.py, and collectible_service.py for improved readability and consistency.

kWAYTV committed:
fix(api_service): update method for Pydantic model serialization

- Replaced `collectible.dict()` with `collectible.model_dump()` for improved serialization of Pydantic models in the ApiService class.

kWAYTV committed:
style(logging): update logger info formatting for consistency

- Changed logger info statements in test-re.py and parser.py to remove f-string formatting for improved consistency.

kWAYTV committed:
chore(docker): update docker-compose and Dockerfile for improved configuration

- Renamed service and container to use hyphens for consistency.
- Added persistent data storage for medals and API responses in docker-compose.yml.
- Set environment variable for Python unbuffered output and restart policy.
- Installed additional system dependencies in Dockerfile for image processing.
- Updated file copying in Dockerfile to leverage caching and included necessary project files.
- Created data directories for medals and responses.

kWAYTV committed:
style: format import statements and Pydantic fields for consistency

- Reformatted import statements in main.py, parser.py, collectible_service.py for improved readability.
- Updated Pydantic model fields in collectible.py to use default values for better clarity and consistency.

kWAYTV commented:
merging some more linting changes now

kWAYTV committed:
style: format code for improved readability and consistency

- Reformatted multiline strings and function definitions across config.py, main.py, test-re.py, parser.py, collectible.py, api_service.py, collectible_service.py, and image_processor.py for better clarity.
- Updated logger info statements to maintain consistent formatting.

tom-jm69 commented on src/models/collectible.py:
```diff
@@ -0,0 +1,106 @@
+    )
+    image: Optional[HttpUrl] = Field(None, description="URL to the collectible image")
+
+    @validator("id")
```
| Pydantic V1 style `@validator` validators are deprecated. You should migrate to Pydantic V2 style

tom-jm69 commented on src/models/collectible.py:
```diff
@@ -0,0 +1,106 @@
+            raise ValueError("Collectible ID cannot be empty")
+        return v.strip()
+
+    @validator("image", pre=True)
```
| Pydantic V1 style `@validator` validators are deprecated. You should migrate to Pydantic V2 style

tom-jm69 commented on src/models/collectible.py:
```diff
@@ -0,0 +1,106 @@
+    )
+    require_image: bool = Field(True, description="Only include items with image URLs")
+
+    @validator("types")
```
| Pydantic V1 style `@validator` validators are deprecated. You should migrate to Pydantic V2 style

tom-jm69 commented on src/models/collectible.py:
```diff
@@ -0,0 +1,106 @@
+    total_count: int = Field(0, description="Total number of items in the batch")
+    filtered_count: int = Field(0, description="Number of items after filtering")
+
+    @validator("total_count", "filtered_count")
```
| Pydantic V1 style `@validator` validators are deprecated. You should migrate to Pydantic V2 style

kWAYTV committed:
refactor(collectible): update Pydantic validators to field_validator

- Replaced @validator with @field_validator for improved validation handling in the Collectible model.
- Reformatted description field for consistency.
- Updated validation methods to use class methods for better structure.

tom-jm69 commented:
Looks good to me.
Thank you for your contribution!

Maybe we could decrease the docker image size, as it now is over triple the original size (maybe by using multi-stage builds).

kWAYTV committed:
perf(docker): optimize image size with multi-stage build

- Implement multi-stage Dockerfile separating build and runtime deps
- Reduce final image size by 60-70% (from ~300-400MB to 131MB)
- Add security improvements with non-root user (appuser)
- Create optimized .dockerignore to exclude unnecessary files
- Maintain full functionality while dramatically reducing image size

kWAYTV committed:
feat(docker): enhance README and add Docker setup documentation

- Updated README with instructions for using pre-built Docker images and local builds.
- Added new document for Docker Hub setup with GitHub Actions for automated image building and publishing.
- Included details on multi-platform support and troubleshooting tips.

kWAYTV committed:
feat(docker): update Docker setup documentation for tagging and builds

- Revised README and Docker setup documentation to clarify tagging strategy for Docker images.
- Updated workflow triggers to build and push images on any branch push, with specific tagging for the main branch.
- Added examples of branch-specific tags for better understanding.
